src/modules/Packages.rb

```diff
@@ -788,30 +788,29 @@ def ForceFullRepropose
       nil
     end
 
-    # Reset package selection, but keep objects of specified type
-    # @param [Array<Symbol>] keep a list of symbols specifying type of objects to be kept
+    # Reset package selection, but keep the selected objects of the specified type
+    # @param [Array<Symbol>] keep a list of symbols specifying type of objects to be kept selected
     def Reset(keep)
-      keep = deep_copy(keep)
       restore = []
-      Builtins.foreach(keep) do |type|
-        selected = Pkg.ResolvableProperties("", type, "")
-        Builtins.foreach(selected) do |s|
-          restore = Builtins.add(
-            restore,
-            { "type" => type, "name" => Ops.get_string(s, "name", "") }
-          )
+
+      # collect the currently selected resolvables
+      keep.each do |type|
+        resolvables = Pkg.ResolvableProperties("", type, "")
+
+        resolvables.each do |resolvable|
+          # only selected items but ignore the selections done by solver,
+          # during restoration they would be changed to be selected by YaST and they
+          # will be selected by solver again anyway
+          if resolvable["status"] == :selected && resolvable["transact_by"] != :solver
+            restore << { "type" => type, "name" => resolvable["name"] }
+          end
         end
       end
 
-      # This reset keep user-made changes
-      # BNC #446406
+      # This keeps the user-made changes (BNC#446406)
       Pkg.PkgApplReset
-      Builtins.foreach(restore) do |res|
-        Pkg.ResolvableInstall(
-          Ops.get_string(res, "name", ""),
-          Ops.get_symbol(res, "type")
-        )
-      end
+
+      restore.each { |res| Pkg.ResolvableInstall(res["name"], res["type"]) }
```

Comment:
when I see usage of map, I think it is a bit overkill to use hash here. simple pair is enough. and you can name it.
so above you can add `restore << [ type, resolvable["name"]]`
and here `restore.each { |type, name| Pkg.ResolvableInstall(name, type) }`